phoenix-core/src/main/java/org/apache/phoenix/schema/MetaDataClient.java

```diff
@@ -4362,18 +4370,40 @@ public MutationState alterIndex(AlterIndexStatement statement) throws SQLExcepti
                     // Set so that we get the table below with the potentially modified rowKeyOrderOptimizable flag set
                     indexRef.setTable(result.getTable());
                     if (newIndexState == PIndexState.BUILDING && isAsync) {
-                        try {
-                            tableUpsert = connection.prepareStatement(UPDATE_INDEX_REBUILD_ASYNC_STATE);
-                            tableUpsert.setString(1,
-                                    connection.getTenantId() == null ? null : connection.getTenantId().getString());
-                            tableUpsert.setString(2, schemaName);
-                            tableUpsert.setString(3, indexName);
-                            tableUpsert.setLong(4, result.getTable().getTimeStamp());
-                            tableUpsert.execute();
-                            connection.commit();
-                        } finally {
-                            if (tableUpsert != null) {
-                                tableUpsert.close();
+                        if (isAll) {
+                            List<Task.TaskRecord> tasks = Task.queryTaskTable(connection, schemaName, tableName, PTable.TaskType.INDEX_REBUILD,
+                                    tenantId, indexName);
+                            if (tasks == null || tasks.size() == 0) {
+                                Timestamp ts = new Timestamp(EnvironmentEdgeManager.currentTimeMillis());
+                                JsonObject jsonObject = new JsonObject();
+                                jsonObject.addProperty(INDEX_NAME, indexName);
+                                jsonObject.addProperty(REBUILD_ALL, true);
+                                try {
+                                    Task.addTask(connection, PTable.TaskType.INDEX_REBUILD,
+                                            tenantId, schemaName,
+                                            dataTableName, PTable.TaskStatus.CREATED.toString(),
+                                            jsonObject.toString(), null, ts, null, true);
+                                    connection.commit();
+                                } catch (IOException e) {
+                                    throw new SQLException("Exception happened while adding a System.Task" + e.toString());
+                                }
+                            }
+                        } else {
+                            try {
+                                tableUpsert = connection.prepareStatement(UPDATE_INDEX_REBUILD_ASYNC_STATE);
```

Comment:
Would be good to be able to eliminate this else block and just push everything through Task, just with the rebuild all flag set or not set. However, we need to verify that this doesn't cause unexpected side effects or regressions. Started a thread above about that.

Reply:
I thought about that. It will be a huge change with unknown consequences. I think it is a bit risky to move everything without scoping. We already have a Jira for that (PHOENIX-5212). I don't want to completely change how we build indexes with the scope of this jira.

Reply:
Good idea @gjacoby126! But, I'd agree with @gokceni to not do it here. Perhaps another Jira or another PR.

Reply:
@priyankporwal @gokceni - Agree that it can be another JIRA, such as PHOENIX-5212.